percolator 3.05 (new formula)

ebzheng opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/brewsci/homebrew-bio/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/brewsci/homebrew-bio/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source FORMULA`, where `FORMULA` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict FORMULA` (after doing `brew install FORMULA`)?

-----
This is my first formula, so there will probably be issues. I know about these:

- I used a crappy `percolator -h` test, but percolator itself actually includes a set of unit tests that run with `make test` when building it manually -- any recommendations on how to go about adapting those into the homebrew framework? (I'm not _entirely_ sure I've correctly built the optional XML support.)
- As alluded to above, XML support is configurable on build; I've opted to include it for simplicity, but I'm not actually sure it worked?
- For some reason, when building percolator manually from source, I had to do a lot more surgery on the build system (e.g. cmake was picking up a conflicting system `xsd` instead of the brew one) plus some messing with compiler flags. I didn't need to do that with my draft formula, which could be Homebrew Magic, or maybe a sign that the XML support got skipped...?
- The [license](https://github.com/percolator/percolator/blob/master/license.txt) for percolator is a bit weird -- it's primarily Apache 2, with source libraries at BSD and MIT, but `libsvm` has a different license?
- Finally, I accidentally have the merge commit from reupdating my personal fork in this PR...

ebzheng commented:
Hmm -- [this](https://github.com/brewsci/homebrew-bio/pull/1202/checks#step:5:126) suggests that no, I did not include XML support correctly beyond the obvious missing `boost` bit.

ebzheng commented:
So in my test environment (a spare Ubuntu 18.04 AWS instance), I end up with the following once the missing `boost` dependency is added. This is reminiscent of the issues I had originally (where I started having to mess with compiler flags), so I'll troubleshoot locally before pushing again.

<details>
```[100%] Linking CXX executable percolator
cd /tmp/percolator-20201112-15408-16fv165/percolator-rel-3-05/src && /home/brew/.linuxbrew/Cellar/cmake/3.18.4/bin/cmake -E cmake_link_script CMakeFiles/percolator.dir/link.txt --verbose=1
/home/brew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/g++-7  -Wno-deprecated-declarations -D_GLIBCXX_PARALLEL -fopenmp -DNDEBUG -O3 -DNDEBUG -rdynamic CMakeFiles/percolator.dir/main.cpp.o -o percolator  libperclibrary.a blas/libblas.a fido/libfido.a picked_protein/libpicked_protein.a -Wl,-Bstatic -lxerces-c -Wl,-Bdynamic -lstdc++
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlNetAccessor.cpp.o): In function `xercesc_3_2::CurlNetAccessor::~CurlNetAccessor()':
CurlNetAccessor.cpp:(.text+0xda): undefined reference to `curl_global_cleanup'
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlNetAccessor.cpp.o): In function `xercesc_3_2::CurlNetAccessor::initCurl()':
CurlNetAccessor.cpp:(.text+0x26): undefined reference to `curl_global_init'
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlNetAccessor.cpp.o): In function `xercesc_3_2::CurlNetAccessor::cleanupCurl()':
CurlNetAccessor.cpp:(.text+0x51): undefined reference to `curl_global_cleanup'
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlNetAccessor.cpp.o): In function `xercesc_3_2::CurlNetAccessor::~CurlNetAccessor()':
CurlNetAccessor.cpp:(.text+0x91): undefined reference to `curl_global_cleanup'
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlNetAccessor.cpp.o): In function `xercesc_3_2::CurlNetAccessor::CurlNetAccessor()':
CurlNetAccessor.cpp:(.text+0x176): undefined reference to `curl_global_init'
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlURLInputStream.cpp.o): In function `xercesc_3_2::CurlURLInputStream::readMore(int*)':
CurlURLInputStream.cpp:(.text+0x1a7): undefined reference to `curl_multi_perform'
CurlURLInputStream.cpp:(.text+0x1bd): undefined reference to `curl_multi_info_read'
CurlURLInputStream.cpp:(.text+0x4fc): undefined reference to `curl_multi_fdset'
CurlURLInputStream.cpp:(.text+0x513): undefined reference to `curl_multi_timeout'
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlURLInputStream.cpp.o): In function `xercesc_3_2::CurlURLInputStream::CurlURLInputStream(xercesc_3_2::XMLURL const&, xercesc_3_2::XMLNetHTTPInfo const*)':
CurlURLInputStream.cpp:(.text+0x87a): undefined reference to `curl_multi_init'
CurlURLInputStream.cpp:(.text+0x883): undefined reference to `curl_easy_init'
CurlURLInputStream.cpp:(.text+0x8ca): undefined reference to `curl_easy_setopt'
CurlURLInputStream.cpp:(.text+0x8dd): undefined reference to `curl_easy_setopt'
CurlURLInputStream.cpp:(.text+0x8f4): undefined reference to `curl_easy_setopt'
CurlURLInputStream.cpp:(.text+0x909): undefined reference to `curl_easy_setopt'
CurlURLInputStream.cpp:(.text+0x91e): undefined reference to `curl_easy_setopt'
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlURLInputStream.cpp.o):CurlURLInputStream.cpp:(.text+0xa23): more undefined references to `curl_easy_setopt' follow
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlURLInputStream.cpp.o): In function `xercesc_3_2::CurlURLInputStream::CurlURLInputStream(xercesc_3_2::XMLURL const&, xercesc_3_2::XMLNetHTTPInfo const*)':
CurlURLInputStream.cpp:(.text+0xb31): undefined reference to `curl_slist_append'
CurlURLInputStream.cpp:(.text+0xb70): undefined reference to `curl_easy_setopt'
CurlURLInputStream.cpp:(.text+0xba1): undefined reference to `curl_easy_setopt'
CurlURLInputStream.cpp:(.text+0xbb8): undefined reference to `curl_easy_setopt'
CurlURLInputStream.cpp:(.text+0xbcf): undefined reference to `curl_easy_setopt'
CurlURLInputStream.cpp:(.text+0xbdf): undefined reference to `curl_multi_add_handle'
CurlURLInputStream.cpp:(.text+0xc2e): undefined reference to `curl_easy_getinfo'
CurlURLInputStream.cpp:(.text+0xcd1): undefined reference to `curl_easy_setopt'
CurlURLInputStream.cpp:(.text+0xd11): undefined reference to `curl_easy_setopt'
/home/brew/.linuxbrew/lib/libxerces-c.a(CurlURLInputStream.cpp.o): In function `xercesc_3_2::CurlURLInputStream::~CurlURLInputStream()':
CurlURLInputStream.cpp:(.text+0xf07): undefined reference to `curl_multi_remove_handle'
CurlURLInputStream.cpp:(.text+0xf10): undefined reference to `curl_easy_cleanup'
CurlURLInputStream.cpp:(.text+0xf19): undefined reference to `curl_multi_cleanup'
CurlURLInputStream.cpp:(.text+0xf3d): undefined reference to `curl_slist_free_all'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTransService::upperCase(char16_t*)':
ICUTransService.cpp:(.text+0x146): undefined reference to `u_toupper_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTransService::lowerCase(char16_t*)':
ICUTransService.cpp:(.text+0x156): undefined reference to `u_tolower_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTransService::compareIString(char16_t const*, char16_t const*)':
ICUTransService.cpp:(.text+0x1dd): undefined reference to `u_foldCase_60'
ICUTransService.cpp:(.text+0x1e9): undefined reference to `u_foldCase_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTransService::compareNIString(char16_t const*, char16_t const*, unsigned long)':
ICUTransService.cpp:(.text+0x2c8): undefined reference to `u_foldCase_60'
ICUTransService.cpp:(.text+0x2d5): undefined reference to `u_foldCase_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTranscoder::~ICUTranscoder()':
ICUTransService.cpp:(.text+0x3e6): undefined reference to `ucnv_close_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTranscoder::canTranscodeTo(unsigned int)':
ICUTransService.cpp:(.text+0x470): undefined reference to `UCNV_FROM_U_CALLBACK_STOP_60'
ICUTransService.cpp:(.text+0x48e): undefined reference to `ucnv_setFromUCallBack_60'
ICUTransService.cpp:(.text+0x4cf): undefined reference to `ucnv_fromUnicode_60'
ICUTransService.cpp:(.text+0x507): undefined reference to `ucnv_setFromUCallBack_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTranscoder::transcodeTo(char16_t const*, unsigned long, unsigned char*, unsigned long, unsigned long&, xercesc_3_2::XMLTranscoder::UnRepOpts)':
ICUTransService.cpp:(.text+0x5d6): undefined reference to `UCNV_FROM_U_CALLBACK_STOP_60'
ICUTransService.cpp:(.text+0x603): undefined reference to `ucnv_setFromUCallBack_60'
ICUTransService.cpp:(.text+0x648): undefined reference to `ucnv_fromUnicode_60'
ICUTransService.cpp:(.text+0x698): undefined reference to `ucnv_setFromUCallBack_60'
ICUTransService.cpp:(.text+0x6f3): undefined reference to `UCNV_FROM_U_CALLBACK_SUBSTITUTE_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICULCPTranscoder::~ICULCPTranscoder()':
ICUTransService.cpp:(.text+0x7e8): undefined reference to `ucnv_close_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICULCPTranscoder::calcRequiredSize(char16_t const*, xercesc_3_2::MemoryManager*)':
ICUTransService.cpp:(.text+0x883): undefined reference to `ucnv_fromUChars_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICULCPTranscoder::transcode(char16_t const*, char*, unsigned long, xercesc_3_2::MemoryManager*)':
ICUTransService.cpp:(.text+0x99f): undefined reference to `ucnv_fromUChars_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICULCPTranscoder::transcode(char16_t const*, xercesc_3_2::MemoryManager*)':
ICUTransService.cpp:(.text+0xadd): undefined reference to `ucnv_fromUChars_60'
ICUTransService.cpp:(.text+0xbe6): undefined reference to `ucnv_fromUChars_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICULCPTranscoder::calcRequiredSize(char const*, xercesc_3_2::MemoryManager*)':
ICUTransService.cpp:(.text+0xc98): undefined reference to `ucnv_toUChars_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICULCPTranscoder::transcode(char const*, xercesc_3_2::MemoryManager*)':
ICUTransService.cpp:(.text+0xda3): undefined reference to `ucnv_toUChars_60'
ICUTransService.cpp:(.text+0xdee): undefined reference to `ucnv_toUChars_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICULCPTranscoder::transcode(char const*, char16_t*, unsigned long, xercesc_3_2::MemoryManager*)':
ICUTransService.cpp:(.text+0xf29): undefined reference to `ucnv_toUChars_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTranscoder::transcodeFrom(unsigned char const*, unsigned long, char16_t*, unsigned long, unsigned long&, unsigned char*)':
ICUTransService.cpp:(.text+0xfd2): undefined reference to `ucnv_toUnicode_60'
ICUTransService.cpp:(.text+0x1175): undefined reference to `ucnv_getMaxCharSize_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTranscoder::ICUTranscoder(char16_t const*, UConverter*, unsigned long, xercesc_3_2::MemoryManager*)':
ICUTransService.cpp:(.text+0x124e): undefined reference to `ucnv_getMaxCharSize_60'
ICUTransService.cpp:(.text+0x1259): undefined reference to `ucnv_getMinCharSize_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTransService::makeNewXMLTranscoder(char16_t const*, xercesc_3_2::XMLTransService::Codes&, unsigned long, xercesc_3_2::MemoryManager*)':
ICUTransService.cpp:(.text+0x14b0): undefined reference to `ucnv_openU_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(ICUTransService.cpp.o): In function `xercesc_3_2::ICUTransService::makeNewLCPTranscoder(xercesc_3_2::MemoryManager*)':
ICUTransService.cpp:(.text+0x16eb): undefined reference to `ucnv_open_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(XMLUniCharacter.cpp.o): In function `xercesc_3_2::XMLUniCharacter::getType(char16_t)':
XMLUniCharacter.cpp:(.text+0x8): undefined reference to `u_charType_60'
/home/brew/.linuxbrew/lib/libxerces-c.a(RangeToken.cpp.o): In function `xercesc_3_2::RangeToken::getCaseInsensitiveToken(xercesc_3_2::TokenFactory*)':
RangeToken.cpp:(.text+0x108c): undefined reference to `uset_openPatternOptions_60'
RangeToken.cpp:(.text+0x10af): undefined reference to `uset_serialize_60'
RangeToken.cpp:(.text+0x10ed): undefined reference to `uset_serialize_60'
RangeToken.cpp:(.text+0x1100): undefined reference to `uset_getSerializedSet_60'
RangeToken.cpp:(.text+0x1108): undefined reference to `uset_getSerializedRangeCount_60'
RangeToken.cpp:(.text+0x113d): undefined reference to `uset_getSerializedRange_60'
RangeToken.cpp:(.text+0x1163): undefined reference to `uset_setSerializedToOne_60'
RangeToken.cpp:(.text+0x116b): undefined reference to `uset_close_60'
collect2: error: ld returned 1 exit status
```
</details>

ebzheng commented:
Very strange. The current [formula ](https://github.com/ebzheng/homebrew-bio/pull/2/) builds fine in my actual production environment (a CentOS HPC environment where I'm installing everything to a personal Homebrew directory), but fails both on the Ubuntu 18.04 instance and on [the Github CI](https://github.com/ebzheng/homebrew-bio/pull/2/checks?check_run_id=1392697457)...

I'm somewhat bemused. Any suggestions on what to do for further inspection?

sjackman commented:
```
CMake Error at /home/linuxbrew/.linuxbrew/Cellar/cmake/3.18.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find Boost (missing: Boost_INCLUDE_DIR) (Required is at least
  version "1.46.0")
```
https://github.com/brewsci/homebrew-bio/pull/1202/checks?check_run_id=1392534274#step:5:147

Try adding `depends_on "boost"`. You may need to also set `Boost_INCLUDE_DIR` to point to `Formula["boost"].include`.

Once you've managed to build it locally, run `brew linkage percolator`. If there is no run-time linkage to Boost because it uses only the Boost headers, you can change `depends_on "boost"` to `depends_on "boost" => :build`.

sjackman commented:
> I used a crappy percolator -h test, but percolator itself actually includes a set of unit tests that run with make test when building it manually -- any recommendations on how to go about adapting those into the homebrew framework? (I'm not entirely sure I've correctly built the optional XML support.)

We don't generally run the unit tests, but do run something like an integration test. That is, run the command line program on a super tiny data set. Failing that, you can run something like…
```ruby
assert_match "Usage", shell_output("#{bin}/percolator -h")
```

sjackman commented:
> Finally, I accidentally have the merge commit from reupdating my personal fork in this PR...

You can rebase your PR onto the `develop branch` with `git rebase -i origin/develop`. You can also squash your commits down to a single commit while you're at it, but you don't have to. I'll also squash it when I merge the PR.

sjackman commented:
> The license for percolator is a bit weird -- it's primarily Apache 2, with source libraries at BSD and MIT, but libsvm has a different license?

See https://docs.brew.sh/License-Guidelines

You want something like…
```ruby
license all_of: ["Apache-2.0", "BSD-2-Clause", "MIT"]
```

ebzheng commented:
@sjackman @SeekingMeaning -- Thanks for your help; I've addressed your points, but the builds still fail. It's very strange to me -- it _looks_ like the linkage failures are actually coming from `xerces-c`. I'm actually wondering if there's something funny with that formula?

Along those lines, since I'm technically running a non-bottled installation, I tried a fresh theoretically-vanilla brand-new Homebrew-on-Linux in `/home/linuxbrew/.linuxbrew/`. 
`brew install xerces-c` gives the following. Should it be emitting "Reinstalling 1 broken dependent" immediately?

<details>

```
...
==> Installing xerces-c
==> cmake ..
==> make
==> ctest -V
==> make install
==> make clean
==> cmake .. -DBUILD_SHARED_LIBS=OFF
==> make
🍺  /home/linuxbrew/.linuxbrew/Cellar/xerces-c/3.2.3_1: 1,684 files, 43.5MB, built in 9 minutes 35 seconds
==> No outdated dependents to upgrade!
==> Checking for dependents of upgraded formulae...
==> Reinstalling 1 broken dependent from source:
xerces-c
==> Caveats
==> openssl@1.1
A CA file has been bootstrapped using certificates from the system
keychain. To add additional certificates, place .pem files in
  /home/linuxbrew/.linuxbrew/etc/openssl@1.1/certs

and run
  /home/linuxbrew/.linuxbrew/opt/openssl@1.1/bin/c_rehash
==> cmake
Emacs Lisp files have been installed to:
  /home/linuxbrew/.linuxbrew/share/emacs/site-lisp/cmake
```

</details>

Moreover, `brew linkage xerces-c` shows that it's linking against system `icuuc` without having a dependency (which was probably causing at least one of the problems I had on CentOS 7). That said, I don't think I see any improvement between `xerces-c` linked against system `icuuc` vs. `xerces-c` linked against Homebrew `icuuc`.

<details>

```
(base) ubuntu@system:~$ brew linkage xerces-c
System libraries:
  /lib/x86_64-linux-gnu/libc.so.6
  /lib/x86_64-linux-gnu/libgcc_s.so.1
  /lib/x86_64-linux-gnu/libm.so.6
  /lib/x86_64-linux-gnu/libpthread.so.0
  /usr/lib/x86_64-linux-gnu/libicuuc.so.60
  /usr/lib/x86_64-linux-gnu/libstdc++.so.6
  libxerces-c-3.2.so
Homebrew libraries:
  /home/linuxbrew/.linuxbrew/lib/libcurl.so.4 (curl)
Unwanted system libraries:
  /usr/lib/x86_64-linux-gnu/libicuuc.so.60
  libxerces-c-3.2.so
```

</details>

I'll keep investigating, but if please let me know if I'm barking up the wrong tree...

ebzheng commented:
Hmm. If I'm reading the logs correctly, it looks like depending on Homebrew `gcc` was the key! (Not sure what's up with the Mac failure though...)

> We don't generally run the unit tests, but do run something like an integration test. That is, run the command line program on a super tiny data set.

I probably didn't use quite the best language -- the built-in testing suite appears to be a combination of lower-level I/O plus tiny-dataset correctness checking. From some preliminary poking at it, it looks like `ctest` drives a python script that does the actual testing and checking, so I might be able to extract some key bits into the Homebrew framework. (Speaking of which -- where is the testing framework documented? I don't fully understand I can / can't access within the test environment...)

sjackman commented:
> Speaking of which -- where is the testing framework documented? I don't fully understand I can / can't access within the test environment...

https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula
https://rubydoc.brew.sh/Homebrew/Assertions.html

I'm not sure if this is exactly the right documentation for the `assert_*` family of functions.
https://docs.ruby-lang.org/en/2.0.0/MiniTest/Assertions.html

sjackman commented:
```console
$ brew linkage xerces-c
System libraries:
  /lib/x86_64-linux-gnu/libc.so.6
  /lib/x86_64-linux-gnu/libgcc_s.so.1
  /lib/x86_64-linux-gnu/libm.so.6
  /lib/x86_64-linux-gnu/libpthread.so.0
  /lib/x86_64-linux-gnu/libstdc++.so.6
Homebrew libraries:
  /home/linuxbrew/.linuxbrew/lib/libcurl.so.4 (curl)
  /home/linuxbrew/.linuxbrew/Cellar/xerces-c/3.2.3_1/lib/libxerces-c-3.2.so (xerces-c)
```
`xerces-c` didn't have a bottle for Linux for some reason, so it was building from source. Try again now using the bottle.

sjackman commented:
Do you have access to a macOS machine where you're able to reproduce this issue locally? Are you able to reproduce this issue locally outside of Homebrew?

ebzheng commented:
Unfortunately not -- I have effectively no experience with Macs! Just Linux flavors...

<details>

```
Install the project...
/usr/local/Cellar/cmake/3.19.0_1/bin/cmake -P cmake_install.cmake
-- Install configuration: "Release"
CMake Error at cmake_install.cmake:41 (file):
  file cannot create directory: /usr/local/share/xml/percolator/xml-pin-1-3.
  Maybe need administrative privileges.
```

</details>

Do Macs do something strange with respect to permissions on `/usr/local/` and subdirectories? I would have thought not since that appears to be the Homebrew directory... Otherwise, I guess it's possible that percolator's cmake script is getting configured incorrectly only on Macs.

sjackman commented on Formula/percolator.rb:
```diff
@@ -0,0 +1,41 @@
+    depends_on "tokyo-cabinet"
+    depends_on "lbzip2"
+    depends_on "pbzip2"
+    depends_on "lzlib"
```
Why are these dependencies needed only on macOS?

ebzheng replied:
I'm not sure -- I'm drawing from the [build script](https://github.com/percolator/percolator/blob/master/admin/builders/osx64_build.sh) found in the repo. (Which incidentally uses brew to install some deps, but not all of them)

sjackman commented:
> Do Macs do something strange with respect to permissions on /usr/local/ and subdirectories? I would have thought not since that appears to be the Homebrew directory... Otherwise, I guess it's possible that percolator's cmake script is getting configured incorrectly only on Macs.

https://github.com/percolator/percolator/blob/6c7cc82ccec074ad814d8a99fa6e21f5bb9508f7/CPack.txt#L22

It fails due to this upstream bug. 👆

sjackman commented:
I've pushed some fixes to your branch. Take a look.

sjackman commented:
CI is 💚 @ebzheng I'll merge this PR if you're happy with it.

ebzheng commented on Formula/percolator.rb:
```diff
@@ -3,39 +3,22 @@ class Percolator < Formula
   depends_on "boost" => :build
   depends_on "cmake" => :build
-  depends_on "gcc" => :build
-  depends_on "xerces-c" => :build
```
Since I'm 99% sure percolator depends on `xerces-c`, how does bottling work without listing it as a dependency?

ebzheng replied:
Nvmd -- I actually realized what happened. When you cleaned up the cmake call, `-DXML_SUPPORT=ON` got lost, which disables the XML support (and hence the dependencies on `xerces-c`, `xsd`, etc.)

sjackman replied:
Feel free to push a commit that adds support for XML back in.

sjackman replied:
It's always nice to start from the solid foundation of a working 🍏 PR.

ebzheng commented:
@sjackman -- is the `xerces-c` bottle broken again? At [`f1dcf3e`](https://github.com/brewsci/homebrew-bio/pull/1202/checks?check_run_id=1444975557) the Linux build took ~5 minutes, and the failing Linux builds are now taking 15. From my local testing, building xerces-c from source is about 10 minutes, and the current Linux build failures look like they're coming from xerces:

<details>

```
cd /tmp/percolator-20201130-17793-byly2q/percolator-rel-3-05/build/src && /home/linuxbrew/.linuxbrew/Cellar/cmake/3.19.1/bin/cmake -E cmake_link_script CMakeFiles/percolator.dir/link.txt --verbose=1
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/g++-5 -lcurl -std=c++11 -Wno-deprecated-declarations -D_GLIBCXX_PARALLEL -fopenmp -DNDEBUG -O3 -DNDEBUG -rdynamic CMakeFiles/percolator.dir/main.cpp.o -o percolator  libperclibrary.a blas/libblas.a fido/libfido.a picked_protein/libpicked_protein.a -Wl,-Bstatic -lxerces-c -Wl,-Bdynamic -lstdc++ 
/home/linuxbrew/.linuxbrew/lib/libxerces-c.a(CurlNetAccessor.cpp.o): In function `xercesc_3_2::CurlNetAccessor::CurlNetAccessor()':
CurlNetAccessor.cpp:(.text+0x11a): undefined reference to `curl_global_init'
/home/linuxbrew/.linuxbrew/lib/libxerces-c.a(CurlNetAccessor.cpp.o): In function `xercesc_3_2::CurlNetAccessor::initCurl()':
CurlNetAccessor.cpp:(.text+0x1e): undefined reference to `curl_global_init'
/home/linuxbrew/.linuxbrew/lib/libxerces-c.a(CurlNetAccessor.cpp.o): In function `xercesc_3_2::CurlNetAccessor::cleanupCurl()':
CurlNetAccessor.cpp:(.text+0x48): undefined reference to `curl_global_cleanup'
...
```

</details>

sjackman commented on Formula/percolator.rb:
```diff
@@ -0,0 +1,35 @@
+  depends_on "cmake" => :build
+  depends_on "xerces-c" => :build
+  depends_on "xsd" => :build
+  depends_on "curl"
```
Is `curl` really needed as a dependency? Does `brew linkage percolator` report run-time library linkage to `curl`?

ebzheng replied:
I believe so?

`brew linkage percolator` on my Ubuntu test environment, at `1d1ca6e`: (Admittedly this is including the `-lcurl` that I added to try and fix the CI curl failures)

<details>

```
brew linkage percolator
System libraries:
  /lib/x86_64-linux-gnu/libc.so.6
  /lib/x86_64-linux-gnu/libm.so.6
  /lib/x86_64-linux-gnu/libpthread.so.0
Homebrew libraries:
  /home/linuxbrew/.linuxbrew/lib/libcurl.so.4 (curl)
  /home/linuxbrew/.linuxbrew/lib/libgcc_s.so.1 (gcc)
  /home/linuxbrew/.linuxbrew/lib/libgomp.so.1 (gcc)
  /home/linuxbrew/.linuxbrew/lib/libstdc++.so.6 (gcc)
```
</details>

sjackman replied:
Okay! Thanks for checking.

sjackman commented on Formula/percolator.rb:
```diff
@@ -0,0 +1,35 @@
+
+    if OS.mac?
+      inreplace "CommonCMake.txt", /(?<=if\(UNIX OR APPLE OR MINGW\).)/m,
+                "set(CMAKE_EXE_LINKER_FLAGS='-framework CoreServices -framework CoreFoundation')"
```
Is this hack needed, and what's the error message without it? It's helpful to include a comment above that says
`# Fix Error: foobar does not frobnicate`
where the error message is copied verbatim from the failed build, which helps with Googling that error message.

ebzheng replied:
This is a WIP -- I don't have a Mac system to test against, but this is targeted against build errors on Mac:

[Example](https://github.com/brewsci/homebrew-bio/runs/1476548379?check_suite_focus=true#step:4:374):

```
Undefined symbols for architecture x86_64:
  "_CFRelease", referenced from:
      xercesc_3_2::MacOSUnicodeConverter::upperCase(char16_t*) in libxerces-c.a(MacOSUnicodeConverter.cpp.o)
      xercesc_3_2::MacOSUnicodeConverter::lowerCase(char16_t*) in libxerces-c.a(MacOSUnicodeConverter.cpp.o)
```

sjackman commented on Formula/percolator.rb:
```diff
@@ -0,0 +1,35 @@
+    mkdir "build" do
+      system "cmake", "..", *std_cmake_args, "-DXML_SUPPORT=ON",
+             "-DCMAKE_CXX_FLAGS='-lcurl -std=c++11'",
+             "-DCMAKE_EXE_LINKER_FLAGS='-lcurl'"
```
Ditto above. Is this hack really needed, and what error message does it fix?

sjackman commented:
For the build failures that you're seeing and hacking around, can you please open bug reports on the upstream `percolator` GitHub repo, and link to those bug reports here?

sjackman commented:
I'd probably prefer to merge `percolator` without XML support for now and eliminate the hacks needed to build it with XML support. We can always open a second PR to add XML support, preferably after the build errors are fixed upstream.

ebzheng commented:
> I'd probably prefer to merge `percolator` without XML support for now and eliminate the hacks needed to build it with XML support. We can always open a second PR to add XML support, preferably after the build errors are fixed upstream.

I concur -- I can reproducibly build `percolator` locally, but that includes _even more_ hacks that I don't necessarily want to introduce to upstream. At least without XML it seems reasonably well-behaved?

sjackman commented:
> At least without XML it seems reasonably well-behaved?

Yeah, it was, except for that `/usr/local` hack. Want to revert this PR back to the last successful point? https://github.com/brewsci/homebrew-bio/pull/1202/commits/de0ff0edbc5a475787cd37cd4cfae2bc38d0320f

ebzheng commented:
Sorry I haven't had time to work on this until now. I rolled back to de0ff0e and added a caveats block to note the lack of XML support (which is unfortunate but probably survivable for most uses). Resolution is easiest on macOS given that `percolator` has an official installation script and most users probably have sudo access, so I wrote in a reference to their repo. I'm not sure what to do for Linux users, since I anticipate most linuxbrew users will not have sudo access. (If they do, the major package managers already have a package built...)

It looks to me like the failure on macOS is due to something going wrong in one of the gems for the style checks and not an issue in the formula itself.

Finally, I'll try to open the upstream issues to ask about the build stuff in this coming week.

sjackman commented:
Merged. Thank you for your contribution to Brewsci/bio, Eric! 🥇
